libraries/standard/mqtt/src/mqtt_state.c

```diff
@@ -323,23 +546,242 @@ static MQTTStatus_t addRecord( MQTTPubAckInfo_t * records,
+
+    /* Record to be deleted if the state transition is completed or if a PUBREC
+     * is received for an outgoing QoS2 publish. When a PUBREC is received,
+     * record is deleted and added back to the end of the records to maintain
+     * ordering for PUBRELs. */
+    shouldDeleteRecord = ( ( newState == MQTTPublishDone ) || ( newState == MQTTPubRelSend ) ) ? true : false;
+    isTransitionValid = validateTransitionAck( currentState, newState );
+
+    if( isTransitionValid == true )
+    {
+        status = MQTTSuccess;
+
+        /* Update record for acks. When sending or receiving acks for packets that
+         * are resent during a session reestablishment, the new state and
+         * current state can be the same. No update of record required in that case. */
+        if( currentState != newState )
+        {
+            updateRecord( records,
+                          recordIndex,
+                          newState,
+                          shouldDeleteRecord );
+
+            /* For QoS2 messages, in order to preserve the message ordering, when
+             * a PUBREC is received for an outgoing publish, the record should be
+             * moved to the last. This move will help preserve the order in which
+             * a PUBREL needs to be resent in case of a session reestablishment. */
+            if( newState == MQTTPubRelSend )
+            {
+                status = addRecord( records,
+                                    MQTT_STATE_ARRAY_MAX_COUNT,
+                                    packetId,
+                                    MQTTQoS2,
+                                    MQTTPubRelSend );
+            }
+        }
+    }
+    else
+    {
+        /* Invalid state transition. Update the new state to #MQTTStateNull. */
+        *pNewState = MQTTStateNull;
+        LogError( ( "Invalid transition from state %s to state %s.",
+                    MQTT_State_strerror( currentState ),
+                    MQTT_State_strerror( newState ) ) );
+    }
+
+    return status;
+}
+
+/*-----------------------------------------------------------*/
+
+static MQTTStatus_t updateStatePublish( MQTTContext_t * pMqttContext,
+                                        size_t recordIndex,
+                                        uint16_t packetId,
+                                        MQTTStateOperation_t opType,
+                                        MQTTQoS_t qos,
+                                        MQTTPublishState_t currentState,
+                                        MQTTPublishState_t * pNewState )
+{
+    MQTTStatus_t status = MQTTSuccess;
+    MQTTPublishState_t newState = MQTTStateNull;
+    bool isTransitionValid = false;
+
+    assert( pMqttContext != NULL );
+    assert( packetId != MQTT_PACKET_ID_INVALID );
+    assert( qos != MQTTQoS0 );
+    assert( pNewState != NULL );
+
+    newState = *pNewState;
+    isTransitionValid = validateTransitionPublish( currentState, newState, opType, qos );
```

Comment:
Can you add a comment about how this will always succeed if opType is a receive? This is due to the fact that the passed in currentState must be MQTTStateNull, since MQTT_UpdateStatePublish does not perform a lookup for receives.